### PR TITLE
ssb-patchwork: fix GTK file selection dialog

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

Fix a bug with file selection dialogs.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ping @asymmetric

### CODE REVIEW

**Comment by asymmetric** [comment]
Works like a charm! :heart:

File: pkgs/applications/networking/ssb/patchwork/default.nix [hunk 1]
```diff
@@ -29,16 +30,20 @@ let
     categories = "Network;";
   };
 
-in
-  symlinkJoin {
-    inherit name;
-    paths = [ binary ];
+in stdenvNoCC.mkDerivation {
```
Comment (worldofpeace) [inline pkgs/applications/networking/ssb/patchwork/default.nix]: I don't really get why we have to use the `NoCC` version here? It has little benefit.

File: pkgs/applications/networking/ssb/patchwork/default.nix [hunk 2]
```diff
@@ -29,16 +30,20 @@ let
...
-      mkdir -p $out/share/pixmaps/ $out/share/applications
-      cp ${appimage-contents}/ssb-patchwork.png $out/share/pixmaps
-      cp ${desktopItem}/share/applications/* $out/share/applications/
-    '';
+  nativeBuildInputs = [ wrapGAppsHook ];
+
+  buildCommand = ''
+    makeWrapper "${binary}/bin/ssb-patchwork" "$out/bin/ssb-patchwork" \
+      --prefix XDG_DATA_DIRS : "${gsettings-desktop-schemas}/share/gsettings-schemas/${gsettings-desktop-schemas.name}" \
+      --prefix XDG_DATA_DIRS : "${gtk3}/share/gsettings-schemas/${gtk3.name}" \
```
Comment (worldofpeace) [inline pkgs/applications/networking/ssb/patchwork/default.nix]: this is an improper use of wrapGAppsHook.
Please read https://nixos.org/nixpkgs/manual/#ssec-gnome-hooks, it exists so you don't have to do this.

It also should automatically wrap everything at `$out/bin`. If you make a symlink here for `${binary}/bin/ssb-patchwork $out/bin/ssb-patchwork` you shouldn't have to do anything.

**Comment by worldofpeace** [comment]
We could probably backport this bugfix to 20.03

File: pkgs/applications/networking/ssb/patchwork/default.nix (hunk 2 above)
Comment (ehmry) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: I would prefer an automatic solution but it doesn't work here.

**Comment by flokli** [comment]
This uses a `buildFHSUserEnv` under the hood - I'd expect things to just work out of the box inside one ot these, so wrapGappsHook or other custom wrapping shouldn't be necessary at all.

@jtojnar, @tilpner, could you take a look?

**Comment by worldofpeace** [comment]
> This uses a `buildFHSUserEnv` under the hood - I'd expect things to just work out of the box inside one ot these, so wrapGappsHook or other custom wrapping shouldn't be necessary at all.
> 
> @jtojnar, @tilpner, could you take a look?

Gsettings-schemas are relocated in `share/gsettings-schemas/${name}/glib-2.0`, not the fhs location. I would obviously be very familiar with this :confused:

File: pkgs/applications/networking/ssb/patchwork/default.nix (hunk 2 above)
Comment (worldofpeace) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: Could you explain why? If you can't you can just use the makeWrapper with the arguments wrapGAppsHook collects ` "${gappsWrapperArgs[@]}"`.

**Comment by tilpner** [comment]
> This uses a `buildFHSUserEnv` under the hood - I'd expect things to just work out of the box inside one ot these, so wrapGappsHook or other custom wrapping shouldn't be necessary at all.
> 
> @jtojnar, @tilpner, could you take a look?

I have little knowledge about how gsettings is supposed to work, and what this gschema.compiled is.
Here's how you can play with the FHS env interactively:

```
$ APPIMAGE_DEBUG_EXEC=bash /nix/store/md7rlyjkkzl84pimmwmh3qs986czs1rf-Patchwork-3.17.6/bin/ssb-patchwork
bash-4.4$ find -L /usr/share/gsettings-schemas
/usr/share/gsettings-schemas
/usr/share/gsettings-schemas/gtk+3-3.24.14
/usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0
/usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas
/usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/org.gtk.Settings.Debug.gschema.xml
/usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/org.gtk.Settings.EmojiChooser.gschema.xml
/usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/org.gtk.Demo.gschema.xml
/usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/gschemas.compiled
/usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/org.gtk.Settings.FileChooser.gschema.xml
/usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/org.gtk.Settings.ColorChooser.gschema.xml
bash-4.4$ echo $XDG_DATA_DIRS
/home/tilpner/.nix-profile/share:/etc/profiles/per-user/tilpner/share:/nix/var/nix/profiles/default/share:/run/current-system/sw/share
```
It seems like a version gschemas.compiled is installed, but buildFHSUserEnv might not set XDG_DATA_DIRS at all. If it's supposed to, you can try adding

```
profile = ''
  export XDG_DATA_DIRS=/usr/share/...
'';
```

to https://github.com/NixOS/nixpkgs/blob/afb035a32b18e0d128949a5f10c9fb0a304863a7/pkgs/build-support/appimage/default.nix#L25-L32

**Comment by worldofpeace** [comment]
> > This uses a `buildFHSUserEnv` under the hood - I'd expect things to just work out of the box inside one ot these, so wrapGappsHook or other custom wrapping shouldn't be necessary at all.
> > @jtojnar, @tilpner, could you take a look?
> 
> I have little knowledge about how gsettings is supposed to work, and what this gschema.compiled is.
> Here's how you can play with the FHS env interactively:
> 
> ```
> $ APPIMAGE_DEBUG_EXEC=bash /nix/store/md7rlyjkkzl84pimmwmh3qs986czs1rf-Patchwork-3.17.6/bin/ssb-patchwork
> bash-4.4$ find -L /usr/share/gsettings-schemas
> /usr/share/gsettings-schemas
> /usr/share/gsettings-schemas/gtk+3-3.24.14
> /usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0
> /usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas
> /usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/org.gtk.Settings.Debug.gschema.xml
> /usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/org.gtk.Settings.EmojiChooser.gschema.xml
> /usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/org.gtk.Demo.gschema.xml
> /usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/gschemas.compiled
> /usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/org.gtk.Settings.FileChooser.gschema.xml
> /usr/share/gsettings-schemas/gtk+3-3.24.14/glib-2.0/schemas/org.gtk.Settings.ColorChooser.gschema.xml
> bash-4.4$ echo $XDG_DATA_DIRS
> /home/tilpner/.nix-profile/share:/etc/profiles/per-user/tilpner/share:/nix/var/nix/profiles/default/share:/run/current-system/sw/share
> ```
> 
> It seems like a version of gschemas.compiled is installed, but buildFHSUserEnv might not set XDG_DATA_DIRS at all. If it's supposed to, you can try adding
> 
> ```
> profile = ''
>   export XDG_DATA_DIRS=/usr/share/...
> '';
> ```
> 
> to
> 
> https://github.com/NixOS/nixpkgs/blob/afb035a32b18e0d128949a5f10c9fb0a304863a7/pkgs/build-support/appimage/default.nix#L25-L32

wrapGAppsHook will add the gsettings-schemas directories to XDG_DATA_DIRS via the glib setup-hook https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/glib/setup-hook.sh#L2.
You've confirmed what I've said here, we don't install gsettings-schemas into their default location, so we have to use wrapGAppsHook to include the gsettings-schemas.

**Comment by flokli** [comment]
But didn't this PR state that simply adding `wrapGAppsHook` didn't work?

**Comment by worldofpeace** [comment]
> But didn't this PR state that simply adding `wrapGAppsHook` didn't work?

wrapGAppsHook being added also just automatically wraps. Maybe manual wrapping is needed, but you still need wrapGAppsHook to collect `gappsWrapperArgs`.

**Comment by NinjaTrappeur** [comment]
Stupid question from a GTK noob here: why do we need to wrap the binary
to add these schema in the first place?

The appimageTools.wrapType2 function we're using in this package is in
charge of creating an FHS env containing among other things GTK3.

Ain't this FHS environment supposed to provide the
${gtk3}/share/gsettings-schemas/* schemas? Why are we missing them there?

Is there a particular reason why we rather should wrap this particular
binary instead of adding these schemas directly in the
appimageTools.wrapType2 function, thus making sure we won't end up
facing the same issue in other appImage bindists?

**Comment by worldofpeace** [comment]
> Stupid question from a GTK noob here: why do we need to wrap the binary to add these schema in the first place? The appimageTools.wrapType2 function we're using in this package is in charge of creating an FHS env containing among other things GTK3. Ain't this FHS environment supposed to provide the ${gtk3}/share/gsettings-schemas/* schemas? Why are we missing them there? Is there a particular reason why we rather should wrap this particular binary instead of adding these schemas directly in the appimageTools.wrapType2 function, thus making sure we won't end up facing the same issue in other appImage bindists?

As I've said, in nixos we have a setup-hook that relocates all settings schemas to `share/gsettings-schemas/${name}/glib-2.0/schemas`. The FHS location would have been `usr/share/glib-2.0/schemas`. We can fix this problem by having the FHS chrootenv install them in that location https://github.com/NixOS/nixpkgs/pull/83705. That PR should obsolete this.

**Comment by NinjaTrappeur** [comment]
I just realized I forgot the obvious in my previous message: thanks a
lot ehmry!

This particular bug has been a personal pain for a while. My current
workaround was to fire up patchwork from my shell environment, which was
fixing this problem for a reason I failed to identify.

Anyways, kudos!

**Comment by ehmry** [comment]
While this patch may fix Patchwork in the short-time, I'm closing the PR in favor of the general fix, #83705.

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/161739

**Pull request closed without merging**